modules/tinymce/src/core/test/ts/browser/UndoManagerTest.ts

```diff
@@ -440,6 +440,35 @@ UnitTest.asynctest('browser.tinymce.core.UndoManager', function (success, failur
     ok(editor.isDirty(), 'Dirty state should be true');
   });
 
+  suite.asyncTest('Dirty state on second AddUndo', (editor, done, die) => {
+    editor.setContent('<p>a</p>');
+    LegacyUnit.setSelection(editor, 'p', 1);
+
+    let first = true;
+    const test = function () {
```

Comment:
Something I've seen some of the other tests do in situations like this is make the test function just append the current state to an array, and then assert that the array looks how you'd expect it to at the end.

I find looking at a sequence of events or observations `[{ isDirty: false }, { isDirty: true }]` makes it easier to get a linear grasp of when things are occurring.

Just a suggestion, though, don't feel you have to change it if you can read it better this way.

Reply:
This would probably be a good idea, as then this wouldn't need to be an async test either, as manually firing those keyboard events should ensure the `AddUndo` event has fired before the end of the function.

Reply:
There's some sort of race condition where whenever I try to convert this test into being synchronous, the suite enters a loop. I'll redo the test when I redo the test file next sprint.